File: ants_book/chapter2-1/maze.cpp
//encoding:utf-8
#include <iostream> //c++
#include<time.h> //時間にかんして
#include<stack> //スタックに関して　最新のデータから捨てる
#include<queue> //キューに関して　最古のデータから捨てる

#define ll long long
using namespace std;

const int INF = 10000;

typedef pair<int,int> P;

int n,m;
char field[100][101];
int sx,sy;
int gx,gy;
//最短郷里を保存
int d[100][100];

int dx[4] = {1,0,-1,0};
int dy[4] = {0,1,0,-1};

int bfs(){
  queue<P> que;
  for(int i=0;i<n;i++){
    for(int j=0;j<m;j++){
      d[i][j] = INF;
    }
  }
  que.push(P(sx,sy));
  d[sx][sy] = 0;

  //que が空になるまでループ
  while(que.size()){
    P p = que.front();
    que.pop();

    if(p.first==gx && p.second==gy){
      break;
    }
    for(int i=0;i<4;i++){
      int nx = p.first+dx[i];
      int ny = p.second+dy[i];

      if (0<=nx && nx<n && 0<=ny && ny<m && field[nx][ny]!='#' && d[nx][ny]==INF){
        que.push(P(nx,ny));
        d[nx][ny] = d[p.first][p.second]+1;
      }
    }
  }
  return d[gx][gy];
}

int main(){
  cin>>n>>m;
  cin>>sx>>sy>>gx>>gy;
  for(int i=0;i<n;i++){
    for(int j=0;j<m;j++){

      cin>>field[i][j];
    }

  }
  sy--;
  sx--;
  gx--;
  gy--;
  int ans = bfs();
  cout<<ans<<"\n";
  for(int i=0;i<n;i++){
    for(int j=0;j<m;j++){
      cout<<d[i][j]<<" ";
    }
    cout<<"\n";
  }
  return 0;
}

File: ants_book/chapter2/maze.py
# encoding:utf-8

import copy
import numpy as np
import random
import queue
# que.put() or que.get()


n,m = map(int,input().split())
sx,sy = map(int,input().split())
gx,gy = map(int,input().split())

sx-=1
sy-=1
gx-=1
gy-=1



field = []
inf = 10**2
visited = [[inf for i in range(m)] for i in range(n)]

dx = [0,1,0,-1]
dy = [1,0,-1,0]

for i in range(n):
    field.append(list(str(input())))

q = queue.Queue()


# Breadth-First Serch
def bfs(sx,sy):
    field[sx][sy] = "."
    visited[sx][sy] = 0
    q.put([sx,sy])
    while not q.empty():
        x,y = q.get()
        if x==gx and y==gy:
            break
        for i in range(4):
            nx = x+dx[i]
            ny = y+dy[i]
            if 0<=nx<n and 0<=ny<m and field[nx][ny]=="." and visited[nx][ny]==inf:
                q.put([nx,ny])
                visited[nx][ny] = visited[x][y] + 1

    return visited[gx][gy]

print(bfs(sx,sy))
